include/exadg/incompressible_navier_stokes/spatial_discretization/calculators/wall_shear_stress_calculator.cpp

```diff
@@ -0,0 +1,232 @@
+  // front since orientation of boundary faces *should be* constant
+  unsigned int const    fe_degree = matrix_free->get_dof_handler(dof_index).get_fe(0).degree;
+  dealii::FESystem<dim> fe_system(dealii::FE_DGQ<dim>(fe_degree), dim);
+  dealii::QGaussLobatto<dim - 1> face_quadrature(fe_degree + 1);
+  dealii::FEFaceValues<dim>      fe_face_values(fe_system,
+                                           face_quadrature,
+                                           dealii::update_quadrature_points);
+  dealii::QGaussLobatto<dim>     quadrature(fe_degree + 1);
+  dealii::FEValues<dim>          fe_values(fe_system, quadrature, dealii::update_quadrature_points);
+
+  face_to_cell_index.clear();
+  for(auto const & cell : matrix_free->get_dof_handler(dof_index).active_cell_iterators())
+  {
+    if(cell->is_locally_owned() == false)
+      continue;
+
+    fe_values.reinit(cell);
+    double const tol_sqrd = std::pow((cell->diameter() * rel_tol), 2);
+
+    face_to_cell_index.resize(cell->n_faces());
+
+    for(auto const face : cell->face_indices())
+    {
+      fe_face_values.reinit(cell, face);
+      std::vector<dealii::types::global_dof_index> face_cell_index_matches(
+        fe_face_values.n_quadrature_points, dealii::numbers::invalid_dof_index);
+
+      for(auto const i : fe_face_values.quadrature_point_indices())
+      {
+        dealii::Point<dim> const point_face = fe_face_values.quadrature_point(i);
+        for(auto const j : fe_values.quadrature_point_indices())
+        {
+          if(point_face.distance_square(fe_values.quadrature_point(j)) < tol_sqrd)
+          {
+            face_cell_index_matches[i] = j;
+            break;
+          }
+        }
+      }
+
+      face_to_cell_index[face] = face_cell_index_matches;
+    }
+
+    break;
+  }
+}
+
+template<int dim, typename Number>
+void
+WallShearStressCalculator<dim, Number>::compute_wall_shear_stress(
+  VectorType &                                  dst,
+  VectorType const &                            src,
+  std::shared_ptr<dealii::Mapping<dim> const>   mapping,
+  std::vector<dealii::types::boundary_id> const write_wall_shear_stress_on_IDs) const
+{
+  dst = 0;
+
+  // fill boundary ID to bool map OR write on all boundaries
+  AssertThrow(
+    write_wall_shear_stress_on_IDs.size() > 0,
+    dealii::ExcMessage(
+      "Provide at least write_wall_shear_stress_on_IDs = {dealii::numbers::invalid_boundary_id}.")) bool
+    write_on_all_boundary_IDs = false;
+  if(write_wall_shear_stress_on_IDs.size() == 1)
+  {
+    if(write_wall_shear_stress_on_IDs[0] == dealii::numbers::invalid_boundary_id)
+      write_on_all_boundary_IDs = true;
+  }
+  std::map<dealii::types::boundary_id, bool> write_on_boundary_IDs;
+  for(auto const id : write_wall_shear_stress_on_IDs)
+    write_on_boundary_IDs.insert({id, true});
+
+  unsigned int const dofs_per_component =
+    matrix_free->get_dof_handler(dof_index).get_fe(0).dofs_per_cell / dim;
```

Comment:
Can't deal.II provide the information about scalar dofs per cell directly? Where is this variable actually used?

Reply:
did not fine one in [classFiniteElement](https://www.dealii.org/current/doxygen/deal.II/classFiniteElement.html), and did not get lucky with get_sub_fe(component_mask).
This form here works for all ReferenceCell types, so that should be fine as far as I am concerned?